File: android/app/src/main/java/com/gesture/controller/overlay/PhoneActionExecutor.kt
package com.gesture.controller.overlay

import com.gesture.controller.gesture.Gesture

interface ActionExecutorListener {
    fun onCursorMoved(x: Float, y: Float) {}
    fun onGestureExecuted(gesture: Gesture) {}
    fun onWakeStateChanged(isAwake: Boolean) {}
    fun onFeedbackMessage(msg: String) {}
    fun onSettingsChanged(settings: CursorSettings) {}
}

class PhoneActionExecutor {
    private val listeners = mutableSetOf<ActionExecutorListener>()
    private var handX = 0.5f
    private var handY = 0.5f
    private var isAwake = false

    fun addListener(listener: ActionExecutorListener) {
        listeners.add(listener)
    }

    fun removeListener(listener: ActionExecutorListener) {
        listeners.remove(listener)
    }

    fun updateCursor(x: Float, y: Float) {
        handX = x.coerceIn(0f, 1f)
        handY = y.coerceIn(0f, 1f)
        listeners.forEach { it.onCursorMoved(handX, handY) }
    }

    fun getCursorPosition(): Pair<Float, Float> = Pair(handX, handY)

    fun execute(gesture: Gesture) {
        if (gesture == Gesture.NONE) return
        listeners.forEach {
            it.onGestureExecuted(gesture)
            it.onFeedbackMessage(gesture.name.replace("_", " "))
        }
    }

    fun updateSettings(settings: CursorSettings) {
        listeners.forEach { it.onSettingsChanged(settings) }
    }

    fun showWakeOverlay() {
        isAwake = true
        listeners.forEach {
            it.onWakeStateChanged(true)
            it.onFeedbackMessage("SYSTEM ACTIVE")
        }
    }

    fun hideOverlays() {
        isAwake = false
        listeners.forEach {
            it.onWakeStateChanged(false)
        }
    }

    fun getWakeState(): Boolean = isAwake
}

File: src/overlay/PhoneActionExecutor.ts
import { CursorSettings, Gesture } from '../types';

export interface ActionExecutorListener {
  onCursorMoved?: (x: number, y: number) => void;
  onGestureExecuted?: (gesture: Gesture) => void;
  onWakeStateChanged?: (isAwake: boolean) => void;
  onFeedbackMessage?: (msg: string) => void;
  onSettingsChanged?: (settings: CursorSettings) => void;
}

export class PhoneActionExecutor {
  private listeners = new Set<ActionExecutorListener>();
  private handX = 0.5;
  private handY = 0.5;
  private isAwake = false;

  addListener(listener: ActionExecutorListener): () => void {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  updateCursor(x: number, y: number): void {
    // Clamp to 0..1
    this.handX = Math.max(0, Math.min(1, x));
    this.handY = Math.max(0, Math.min(1, y));

    this.listeners.forEach((l) => l.onCursorMoved?.(this.handX, this.handY));
  }

  getCursorPosition(): { x: number; y: number } {
    return { x: this.handX, y: this.handY };
  }

  execute(gesture: Gesture): void {
    if (gesture === Gesture.NONE) return;

    this.listeners.forEach((l) => {
      l.onGestureExecuted?.(gesture);
      l.onFeedbackMessage?.(gesture);
    });
  }

  updateSettings(settings: CursorSettings): void {
    this.listeners.forEach((l) => l.onSettingsChanged?.(settings));
  }

  showWakeOverlay(): void {
    this.isAwake = true;
    this.listeners.forEach((l) => {
      l.onWakeStateChanged?.(true);
      l.onFeedbackMessage?.('SYSTEM ACTIVE');
    });
  }

  hideOverlays(): void {
    this.isAwake = false;
    this.listeners.forEach((l) => {
      l.onWakeStateChanged?.(false);
    });
  }

  getWakeState(): boolean {
    return this.isAwake;
  }
}
